docker/daemon.go

```diff
@@ -150,16 +154,42 @@ func mainDaemon() {
 	job.Setenv("TlsCert", *flCert)
 	job.Setenv("TlsKey", *flKey)
 	job.SetenvBool("BufferRequests", true)
-	err := job.Run()
+
+	// The serve API job never exits unless an error occurs
+	// We need to start it as a goroutine and wait on it so
+	// daemon doesn't exit
+	serveAPIWait := make(chan struct{})
+	go func() {
+		defer func() {
+			// if API job ever exits (on error or success)
+			// we want to shutdown cleanly and exit the daemon
+			eng.Shutdown()
+			close(serveAPIWait)
+		}()
+
+		if err := job.Run(); err != nil {
+			log.Errorf("Error from ServeAPI: %v", err)
+		}
+	}()
 
 	// Wait for the daemon startup goroutine to finish
 	// This makes sure we can actually cleanly shutdown the daemon
-	log.Infof("waiting for daemon to initialize")
-	<-daemonWait
-	eng.Shutdown()
+	log.Debug("waiting for daemon to initialize")
+	err := <-daemonInitWait
 	if err != nil {
-		// log errors here so the log output looks more consistent
-		log.Fatalf("shutting down daemon due to errors: %v", err)
+		if !eng.IsShutdown() {
```

Comment:
actually, it's protected by a lock, but I guess at this point it doesn't matter whether I call a function to read it, or call Shutdown(), which starts by reading it :)